### PR TITLE
feat(frontend): Derived store to find pending transactions

### PR DESCRIPTION
**Pull request by lmuntaner**
# Motivation

We want to display a warning messange and disable the send form if the modal is for bitcoin and the user has pending transactions.

The bitcoin addresses are unique. Therefore, I wanted to take advantage of that instead of checking selected networks or tokens.

The idea is to create a derived store creator that will return `true`, `false` or `"loading"` depending on the address and whether the bitcoin addresses have been loaded:

* Return `true` when the address is a bitcoin address and the pending transactions array is not empty.
* Return `"loading"` if the addresses or the pending transactions are not loaded. This is trickier because we also need to check the `testnets`. If testnets are not enabled, then the addresses for testnet and regtest won't be loaded.
* Return `false` if the pending transactions are present and empty.

This derived store will be used in `IcSendForm`:

```
let hasPendingTransactionsStore: Readable<boolean | "loading">;
$: hasPendingTransactionsStore = initHasPendingTransactionsStore(source);

//...

{#if $hasPendingTransactionsStore}
  <WarningMessage>{...}</>
```

# Changes

* New derived store creator `initHasPendingTransactions`.

# Tests

* Add tests for the new derived store.

### CODE REVIEW

**Comment by lmuntaner** [comment]
@peterpeterparker @AntonioVentilii-DFINITY please review

This is still in Draft, but I want to confirm the approach before I write the tests and spend time improving the code.

**Comment by peterpeterparker** [comment]
The pending transactions won't be fetched in a web worker? If there are, you can compute what you need there without arming the UI.

**Comment by lmuntaner** [comment]
> The pending transactions won't be fetched in a web worker? If there are, you can compute what you need there without arming the UI.

Pending transactions don't need to be pulled regularly. That's why I don't think it makes sense to put them in a web worker.

Even if they are loaded in a web worker, they can still go into the store I merged in #2635 , right?

Then the component will read from this derived store and render the warning message or not.

**Comment by peterpeterparker** [comment]
> Pending transactions don't need to be pulled regularly.

Are you sure? If I go to my Bitcoin wallet XYZ and send BTC to my Oisy Wallet, does this result in a "Pending Transaction"?

**Comment by lmuntaner** [comment]
> > Pending transactions don't need to be pulled regularly.
> 
> Are you sure? If I go to my Bitcoin wallet XYZ and send BTC to my Oisy Wallet, does this result in a "Pending Transaction"?

Good question, I see the confusion. These are "sent" pending transactions. This is the only moment where we would disable sending.

Shall I rename the store "pendingTransactionsStore" to "btcPendingSentTransactionsStore"?

**Comment by lmuntaner** [comment]
> Ah, sorry for the confusion. Maybe yes, let's rename the store even if verbose.

Shall I do it after this PR? In this PR? Or in another PR and then fix the merge conflicts in this PR?

File: src/frontend/src/btc/derived/has-pending-transactions.derived.ts
Comment (peterpeterparker) [inline src/frontend/src/btc/derived/has-pending-transactions.derived.ts:34]: Not sure if made the choice on purpose so just in case, you can inline/remove the above first check here:

```
if (nonNullish($btcAddressMainnet) && $btcAddressMainnet !== address) {
   return false;
}

return "loading";
```

File: src/frontend/src/btc/derived/has-pending-transactions.derived.ts
Comment (peterpeterparker) [inline src/frontend/src/btc/derived/has-pending-transactions.derived.ts:44]: maybe a dumb idea but, if you test if `address` is actually `nullish` earlier in this function, would you maybe spare some checks?

Just thinking at if there is a way to reduce the numbers of `if`.

**Review by peterpeterparker — COMMENTED** [review]
Looks good. Just wanted two share two ideas that crossed my mind.

**Comment by peterpeterparker** [comment]
> Shall I do it after this PR? In this PR? Or in another PR and then fix the merge conflicts in this PR?

No strong opinion. Your call.

File: src/frontend/src/btc/derived/has-pending-transactions.derived.ts
Comment (lmuntaner) [reply src/frontend/src/btc/derived/has-pending-transactions.derived.ts]: Done! Thanks!

File: src/frontend/src/btc/derived/has-pending-transactions.derived.ts
Comment (lmuntaner) [reply src/frontend/src/btc/derived/has-pending-transactions.derived.ts]: I have refactored the code using helper variables to explain the ocnditionals. I hope it's more readable now.

I was able to combine different conditions within one variable and still make it readable. Therefore, I removed one if statement.

**Comment by lmuntaner** [comment]
> > Shall I do it after this PR? In this PR? Or in another PR and then fix the merge conflicts in this PR?
> 
> No strong opinion. Your call.

I'll refactor both stores in an upcoming PR after this one is merged.

**Comment by lmuntaner** [comment]
@peterpeterparker ready for another review

**Review by peterpeterparker — APPROVED** [review]
LGTm, thx

**Pull request merged**
